src/Npgsql/ConnectorPool.cs

```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
                                 }
                                 else
                                 {
-                                    using (cancellationToken.Register(() => tcs.SetCanceled()))
+                                    using (cancellationToken.Register(s => ((TaskCompletionSource<NpgsqlConnector?>)s).SetCanceled(), tcs))
                                         await tcs.Task;
                                 }
                             }
                             else
                             {
                                 if (timeout.IsSet)
                                 {
-                                    var timeLeft = timeout.TimeLeft;
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    var timeLeft = timeout.TimeLeft.TotalMilliseconds;
+                                    if (timeLeft > int.MaxValue)
+                                        throw new ArgumentOutOfRangeException(nameof(timeout), "timeout.TimeLeft.TotalMilliseconds must be smaller than int.MaxValue.");
```

Comment:
Just a thought, but since `PoolTests` already has little tests for things like `MaxPoolSize < MinPoolSize`, it might be nice to add one for this too.

E.g.:

https://github.com/npgsql/npgsql/blob/d78f3d6c8ca88d6adb6fb26d77bd8d8293269d0c/src/Npgsql/ConnectorPool.cs#L101-L102


https://github.com/npgsql/npgsql/blob/d78f3d6c8ca88d6adb6fb26d77bd8d8293269d0c/test/Npgsql.Tests/PoolTests.cs#L39-L44

Reply:
Same for `PoolSizeLimit <= short.MaxValue`, if you change it to a runtime check.

Reply:
I'm not going to add a test for a timeout that can take 24 days and is artificially constrained by the BCLs choice for a primitive type. 🙃

Reply:
Ope! Yeahhh, I didn't think that one through on my end!